ffs/filcold/filcold.go

```diff
@@ -77,50 +81,80 @@ func (fc *FilCold) Fetch(ctx context.Context, pyCid cid.Cid, piCid *cid.Cid, wad
 	return ffs.FetchInfo{RetrievedMiner: miner, FundsSpent: fundsSpent}, nil
 }
 
+func (fc *FilCold) calculatePieceSize(ctx context.Context, c cid.Cid) (api.DataSize, error) {
+	// Limiting deal size calculation since is very resource intensive.
+	// In the next Lotus version (v1.1.3) a new API will be used which also calculates
+	// CommP, so we can help Lotus avoid recalculating size and CommP when deals are made.
+	select {
+	case fc.semaphDealPrep <- struct{}{}:
+	case <-ctx.Done():
+		return api.DataSize{}, fmt.Errorf("canceled by context")
+	}
+	defer func() { <-fc.semaphDealPrep }()
+	for {
+		if fc.lsm.SyncHeightDiff() < unsyncedThreshold {
+			break
+		}
+		log.Warnf("lotus backpressure from unsynced node")
+		select {
+		case <-ctx.Done():
+			return api.DataSize{}, fmt.Errorf("canceled by context")
+		case <-time.After(time.Minute):
+		}
+	}
+
+	size, err := fc.dm.CalculatePieceSize(ctx, c)
+	if err != nil {
+		return api.DataSize{}, fmt.Errorf("getting cid cummulative size: %s", err)
+	}
+	return size, nil
+}
+
 // Store stores a Cid in Filecoin considering the configuration provided. The Cid is retrieved using
 // the DAGService registered on instance creation. It returns a slice of ProposalCids that were correctly
 // started, and a slice of with Proposal Cids rejected. Returned proposed deals can be tracked
 // with the WaitForDeal API.
 func (fc *FilCold) Store(ctx context.Context, c cid.Cid, cfg ffs.FilConfig) ([]cid.Cid, []ffs.DealError, uint64, error) {
+	fc.l.Log(ctx, "Calculating piece size...")
+	size, err := fc.calculatePieceSize(ctx, c)
+	if err != nil {
+		return nil, nil, 0, fmt.Errorf("getting cid cummulative size: %s", err)
+	}
+	fc.l.Log(ctx, "Calculated piece size is %d MiB.", size.PieceSize/1024/1024)
+	pieceSize := uint64(size.PieceSize)
+
+	if pieceSize < fc.minPieceSize {
+		return nil, nil, 0, fmt.Errorf("Piece size is below allowed minimum %d MiB", fc.minPieceSize/1024/1024)
+	}
```

Comment:
Calculating piece size is moved before query-asking miners, so if the deal size is lower than limit we avoid doing work of doing miner selection.

Apart from performance reasons, we now use the piece size to select miners, so this change was needed anyway.